### ISSUE TITLE
[BUG] ORC writer/reader fail roundtrip test

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#13498]
ORC writer and/or reader failed my specific test (input file attached). The input `f0.parquet` and `f0.orc` are written by Spark from the same random table. I am not sure if this is the reader issue or the writer issue, or both.

Input files: [f1322341552.parquet.tar.gz](https://github.com/rapidsai/cudf/files/11630138/f1322341552.parquet.tar.gz)

### Read `f0.parquet` => write `out.orc` => read `out.orc`: 
```
f0.parquet:
cudf::struct_view<int8_t,int16_t,int32_t,int64_t,float,double,cudf::string_view,bool,cudf::timestamp_D,cudf::timestamp_ns,numeric::decimal32,numeric::decimal64,numeric::decimal128,>

read out.orc:
cudf::struct_view<int8_t,int16_t,int32_t,int64_t,float,double,cudf::string_view,bool,cudf::timestamp_ms,cudf::timestamp_ns,numeric::decimal32,numeric::decimal64,numeric::decimal128,>
```

In summary, the input `cudf::timestamp_D` becomes `cudf::timestamp_ms`.

### Read `f0.orc` => write `out.orc` => read `out.orc`: 
```
read f0.orc:
cudf::struct_view<int8_t,int16_t,int32_t,int64_t,float,double,cudf::string_view,bool,cudf::timestamp_ms,cudf::timestamp_ns,numeric::decimal32,numeric::decimal64,numeric::decimal128,>

read out.orc:
cudf::struct_view<int8_t,int16_t,int32_t,int64_t,float,double,cudf::string_view,bool,cudf::timestamp_ns,cudf::timestamp_ns,numeric::decimal32,numeric::decimal64,numeric::decimal128,>
```

In summary, the input `cudf::timestamp_ms` becomes `cudf::timestamp_ns`.

**Comment by ttnghia** [comment]
Here is the simple C++ code to run:
```
#include <cudf_test/base_fixture.hpp>
#include <cudf_test/column_utilities.hpp>
#include <cudf_test/column_wrapper.hpp>
#include <cudf_test/cudf_gtest.hpp>
#include <cudf_test/io_metadata_utilities.hpp>
#include <cudf_test/iterator_utilities.hpp>
#include <cudf_test/table_utilities.hpp>
#include <cudf_test/type_lists.hpp>

#include <cudf/io/data_sink.hpp>
#include <cudf/io/datasource.hpp>
#include <cudf/io/orc.hpp>
#include <cudf/io/orc_metadata.hpp>
#include <cudf/io/parquet.hpp>

struct MyTest : public cudf::test::BaseFixture {};

TEST_F(MyTest, Test)
{
  std::string base_path = "/home/nghiat/Devel/data/f1322341552.parquet/";

  auto const opts =
    cudf::io::parquet_reader_options::builder(cudf::io::source_info(base_path + "f0.parquet"))
      .build();
  auto const t_parquet = cudf::io::read_parquet(opts);
  printf("num rows: %d, num cols: %d\n", t_parquet.tbl->num_rows(), t_parquet.tbl->num_columns());

  printf("col 0 from parquet: \n");
  auto c0_parquet = cudf::slice(t_parquet.tbl->view().column(0), {0, 1})[0];
  cudf::test::print(c0_parquet);

  cudf::io::orc_reader_options in_opts0 =
    cudf::io::orc_reader_options::builder(cudf::io::source_info{base_path + "f0.orc"})
      .use_index(false);
  auto t_orc = cudf::io::read_orc(in_opts0);
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(t_parquet.tbl->view().column(0), t_orc.tbl->view().column(0));
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(t_parquet.tbl->view().column(1), t_orc.tbl->view().column(1));
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(t_parquet.tbl->view().column(2), t_orc.tbl->view().column(2));

  printf("col 0 from orc: \n");
  auto c0_orc = cudf::slice(t_orc.tbl->view().column(0), {0, 1})[0];
  cudf::test::print(c0_orc);

  cudf::io::orc_writer_options out_opts = cudf::io::orc_writer_options::builder(
    cudf::io::sink_info{base_path + "out.orc"}, t_orc.tbl->view());
  //    cudf::io::sink_info{base_path + "out.orc"},
  //    t_parquet.tbl->view());
  cudf::io::write_orc(out_opts);

  cudf::io::orc_reader_options in_opts =
    cudf::io::orc_reader_options::builder(cudf::io::source_info{base_path + "out.orc"})
      .use_index(false);
  //      .timestamp_type(cudf::data_type{cudf::type_id::TIMESTAMP_MILLISECONDS});

  auto t_final = cudf::io::read_orc(in_opts);

  printf("col 0 from orc roundtrip: \n");
  auto c0_orc_final = cudf::slice(t_final.tbl->view().column(0), {0, 1})[0];
  cudf::test::print(c0_orc_final);

  //  CUDF_TEST_EXPECT_TABLES_EQUAL(t_parquet.tbl->view(), final.tbl->view());
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(t_parquet.tbl->view().column(0), t_final.tbl->view().column(0));
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(t_parquet.tbl->view().column(1), t_final.tbl->view().column(1));
  CUDF_TEST_EXPECT_COLUMNS_EQUAL(t_parquet.tbl->view().column(2), t_final.tbl->view().column(2));
}

```

**Comment by vuule** [comment]
These seem to be in line with the expected behavior:
> the input cudf::timestamp_D becomes cudf::timestamp_ms.

ORC reader uses ms timestamps for dates by default. This behavior can be modified via `enable_use_np_dtypes` option. I'm not sure why only ORC has this option, which effectively just disables `TIMESTAMP_DAYS`.

> the input cudf::timestamp_ms becomes cudf::timestamp_ns

ORC only has nanoseconds timestamps so we default to this type. The `timestamp_type` option can be used to specify the type for all timestamp columns.

In general, we cannot round trip types that exist in libcudf, but not in the file format. There are a few similar issue about Parquet. If we explicitly want to round trip all types, we'd need to write a libcudf into the metadata.

**Comment by vuule** [comment]
Looking further into `enable_use_np_dtypes`:
From https://github.com/rapidsai/cudf/issues/2384 it seems like Spark does expect a DATE and I assume that `enable_use_np_dtypes(false)` is always used there. Python uses the default, presumably for compatibility with numpy. 
We could change the default to what makes more sense on the C++ side.

**Comment by ttnghia** [comment]
Ah, so `use_np_dtypes(false)` actually can write the correct timestamp. Then the test can pass with such option.

Thanks for the investigation. Close this.
